orangecontrib/text/widgets/owpreprocess.py

```diff
@@ -1,70 +1,1365 @@
+            item.setToolTip(pp_def.summary)
+            item.setData(pp_def, DescriptionRole)
+            item.setData(params, ParametersRole)
+
+            model.appendRow(item)
+        return model
+
+    def __activated(self, index):
+        # On pp option double click.
+        # Get the pp item from the list.
+        list_item = self.preprocessors.itemFromIndex(index)
+        # Extract the pp action class instance.
+        pp_action = list_item.data(DescriptionRole)
+
+        # Reject the double click add if duplicate.
+        for i in range(0, self.preprocessormodel.rowCount()):
+            if pp_action.qualname == self.preprocessormodel.index(i, 0).data(DescriptionRole).qualname:
+                return
+
+        item = QStandardItem()
+        item.setData({}, ParametersRole)
+        item.setData(pp_action.title, Qt.DisplayRole)
+        item.setData(pp_action, DescriptionRole)
+        self.preprocessormodel.appendRow([item])
+
+    def storeSpecificSettings(self):
+        self.stored_settings = self.save(self.preprocessormodel)
+        super().storeSpecificSettings()
+
+    def commit(self):
+        # Sync the model into stored_settings on every change commit.
+        self.storeSpecificSettings()
+        if not self._invalidated:
+            self._invalidated = True
+            QApplication.postEvent(self, QEvent(QEvent.User))
+
+    def handleNewSignals(self):
+        self.apply()
+
+    def customEvent(self, event):
+        if event.type() == QEvent.User and self._invalidated:
+            self._invalidated = False
+            self.apply()
 
     def apply(self):
-        # TODO change this to custom stopwords
-        if self.remove_stpwrds:
-            sw = 'english'
-        pp = Preprocessor(incl_punct=self.include_punctuation, trans=self.transformation,
-                          lowercase=self.lowercase, stop_words=sw)
-        self.send(Output.PREPROCESSOR, pp)
\ No newline at end of file
+        self.error(1, "")
+        # Nothing to do, if there is no input corpus.
+        if self.data is None:
+            return
+
+        # Build a preprocessor instance with the selected parameters.
+        preprocessor = self.build_preproc()
+        if not preprocessor:
+            self.error(1, "No tokenizer specified.")
+            return
+
+        # Take the input corpus and append tokens to the metadata.
+        input_data = self.data.documents
+        output_corp = self.data
+        tokens = preprocessor(input_data)
+
+        print(tokens)
+
+        # Update the input corpus with a list of tokens.
+        self.data.tokens = tokens
+        self.send(Output.PP_CORPUS, self.data)
+
+        # Sync the model into stored_settings on every apply.
+        self.storeSpecificSettings()
+
+    def build_preproc(self):
+        self.warning(1, "")
+        # Preprocessor parameters that we require.
+        p = {"tokenizer": None, "lowercase": False, "stop_words": None, "transformation": None}
```

Comment:
The scripting part alredy has the defults set, why do we need to set them once again here?